### ISSUE TITLE
504 Gateway Timeout on badges inside GitHub README

### ISSUE THREAD

**Issue by jcbcn** [badges/shields#4223]
Are you experiencing an issue with...

- [ ] [shields.io](https://shields.io/#/)

:beetle: 
I am receiving 504 (Gateway Timeout) when viewing badges inside GitHub README's over a range of badge types. This is happening with least one badge in the README each time I try.

Possibly linked to https://github.com/badges/shields/issues/1568.

Thanks

**Comment by paulmelnikow** [comment]
Hi! Thanks for the report.

I haven't been able to reproduce this today on the Shields repos or any other repos that I've noticed.

Is this a public repo where you're seeing this happen?

**Comment by jcbcn** [comment]
Yeah, repo is here: https://github.com/pitch-game/pitch-api

I do have 17 in my README which may be the issue (maybe a few too many!). It is usually only happening on the test and coverage ones. Rather than the top three.

I could move them into individual README's but figured this could be a wider issue.

**Comment by paulmelnikow** [comment]
It should work, even with lots of badges…

I wonder if these are timing out because the upstream requests are slow. GitHub cuts things off at 4 seconds, so if the upstream requests take longer than that, the only thing we can do is put the right answer in the cache and then get it right on the next request (e.g. after refresh).

**Comment by paulmelnikow** [comment]
By the way, these are loading for me:

<img width="855" alt="Screen Shot 2019-10-21 at 7 28 36 PM" src="https://user-images.githubusercontent.com/1487036/67228328-4e91c500-f439-11e9-8d7c-5158ab9c736b.png">

**Comment by jcbcn** [comment]
I'm starting to see this on first load of my frontend repo as well. I've just loaded up the README again this evening and am seeing the same behaviour:

![image](https://user-images.githubusercontent.com/1164656/67321099-07720580-f507-11e9-87ff-978396b09854.png)

It seems like it's slowly creeping back in again. After #1568 was fixed I haven't seen a broken badge for ages, but recently in the past week they seem to be more regular. I wonder if the IP ranges have changed which the cache was using?

**Comment by calebcartwright** [comment]
The other thing with both the Azure DevOps test and coverage badges is that we have to make two API calls to AzDO to get the actual data (the first to get the build id of the latest build and the second to get the respective data for that build)

**Comment by jcbcn** [comment]
Ah I see. The strange thing is this has been working flawlessly ever since #1568, just in the last week I've noticed it happening again. Obviously I'm not sure what's degraded, but the because it had been fine for a while, I assumed it might be the caching again.

**Comment by platan** [comment]
There is something strange happening with our servers in the last hours:
https://metrics.shields.io/d/aESRBSjmz/services?orgId=1&from=now-3h&to=now
https://metrics.shields.io/d/g_1B7zhik/prom-client-default-metrics?orgId=1&refresh=10s&from=now-3h&to=now

**Comment by calebcartwright** [comment]
I believe @paulmelnikow ran a deploy at some point today (not sure when), but if there was a deployment during that same window that could potentially explain the spikes.

**Comment by jcbcn** [comment]
Seems to be fine since the rollback. Thanks again for looking into this. Much appreciated

**Comment by paulmelnikow** [comment]
To clarify, I think @jcbcn's initial report may be related to a capacity issue, or else due to upstream services being slow.

While the symptom was the same, the issue that happened yesterday was probably unrelated.

**Comment by paulmelnikow** [comment]
Have a link?

**Comment by paulmelnikow** [comment]
They loaded for me just now.

<img width="931" alt="Screen Shot 2019-10-24 at 10 46 37 PM" src="https://user-images.githubusercontent.com/1487036/67524068-4abba800-f6b0-11e9-93f7-06a0327203df.png">

Seems like this is happening intermittently. I'm curious how much this is related to Azure DevOps and how much it's intermittent server issues.

**Comment by calebcartwright** [comment]
Other than that short window after the last deploy, I haven't noticed anything either (including various Azure DevOps badges that are loading fine for me 🤷‍♂)

Either way though, it's probably worth revisiting the experiments we discussed in #3874 to update our prod environment

**Comment by jcbcn** [comment]
This looks to be resolved now. Thanks

**Comment by paulmelnikow** [comment]
Thanks! Please open a new issue if this recurs.
